src/frame.js

```diff
@@ -0,0 +1,104 @@
+class FrameNotTen {
+  constructor(frameNumber, rollsWithBonus) {
+    this.frameNumber = frameNumber;
+    this.rollsWithBonus = Number(rollsWithBonus);
+    this.frameScore = [];
+    this.frameBonus = [];
+  }
+
+  _addRollAndReturnBonusForPrevFrames(pins) {
+    this.checkForErrors(pins);
+    if (this.frameNumber != 10 || this.frameScore.length == 0)
+      this.frameScore = this.frameScore.concat([parseInt(pins)]);
+    else if (
+      this.frameNumber == 10 ||
+      (this.frameScore.length > 1 && this.frameScore[0] != 10)
+    )
+      this.frameScore = this.frameScore.concat([parseInt(pins)]);
+    if (this.frameScore.length <= 2)
+      return {
+        bonusPrevFrames: this.sendBonusToPreviousFrames(pins),
+        bonusFutureRolls: this.rollsWithBonus,
+      };
+    else
+      return {
+        bonusPrevFrames: [],
+        bonusFutureRolls: 0,
+      };
+  }
+
+  checkForErrors(pins) {
+    if (isNaN(pins)) this.shouldThrowError("argument should be a number");
+    if (pins < 0) this.shouldThrowError("argument should be greater than 0");
+    if (this.frameEnded()) this.shouldThrowError("this frame has eneded");
+    if (this.tooManyPins(pins) && this.frameNumber != 10)
+      this.shouldThrowError("argument + score > 10");
+  }
+
+  frameEnded() {
+    if (this.frameNumber != 10 && this.frameScore.length >= 2) return true;
```

Comment:
clear logic tree here, could it be even clearer if broken down into component functions?